### PR TITLE
Bump roomba

### PR DESCRIPTION
**Pull request by asaba96**
If you request to bump a roomba, the drone will go over to it, follow it until it turns around, and will block the roomba. 

This PR includes misc. fixes and clean up of the RRE. 

Everything works in the sim, except that block roomba needs some tuning and rework.

### CODE REVIEW

**Comment by asaba96** [comment]
Additionally, this PR depends on a PR in msgs and in the sim.

**Comment by amiller27** [comment]
Have you tested this on latest motion?

**Comment by asaba96** [comment]
Yes

**Comment by asaba96** [comment]
Go to and track work fine. Height hold seems to work beautifully. But block is having some issues.

**Comment by amiller27** [comment]
I ask because `ending_radius` is not an attribute of `QuadMoveGoal`, so hit roomba crashes

**Comment by asaba96** [comment]
Fixed that.

File: src/iarc7_abstract/roomba_request_executer.py [hunk 1]
```diff
@@ -159,38 +231,61 @@ def _run(cls, roomba_request, status_callback):
...
-                elif state == RoombaRequestExecuterState.RECOVER_FROM_SUCCESS:
-                    state = RoombaRequestExecuterState.FAILED_RECOVERY
-                else:
-                    state = RoombaRequestExecuterState.RECOVER_FROM_FAILURE
-            else:
+            """state transitioning"""
+
+            # see http://docs.ros.org/api/actionlib_msgs/html/msg/GoalStatus.html
+            # for status IDs
+            if cls._client.get_state() not in [4,5,9]:
```
Comment (amiller27) [inline src/iarc7_abstract/roomba_request_executer.py]: Can you replace the numbers here with variables that say what these are, or just say what those states are in the comment?

Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py]: added comment

**Comment by amiller27** [comment]
And you've tested it now on latest master?

**Comment by asaba96** [comment]
Yes; track roomba is not compatible with the way we have handled the roomba IDs, but its the only task that is that way. We can change every task to use the full frame id like track roomba is now doing if thats how you wanted to handle that now.

**Comment by asaba96** [comment]
I have fixes ready if thats how you wanted to handle roomba ids. I think its cleaner than this chopping we've been doing

**Comment by amiller27** [comment]
Yeah just make it handle the whole thing like track does now

**Comment by asaba96** [comment]
fixed

File: src/iarc7_abstract/roomba_request_executer.py [hunk 2]
```diff
@@ -64,11 +99,16 @@ def init(cls, server_name):
                 server_name,
                 QuadMoveAction)
 
+            roomba_sub = rospy.Subscriber('roomba_full_states',
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py:101]: Is this going to be made available by the multi-roomba filter?

File: src/iarc7_abstract/roomba_request_executer.py [hunk 3]
```diff
@@ -129,19 +172,46 @@ def _run(cls, roomba_request, status_callback):
...
+                    # Sends the goal to the action server.
+                    cls._client.send_goal(goal)
+                    rospy.logwarn("Following Roomba")
+                    track_sent = True
+
+                if cls._roomba_turning:
+                    cls._client.cancel_goal()
+
+                # rate limiting
+                rospy.sleep(.05)
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py]: This should be a rate.sleep()

Comment (aftersomemath) [reply src/iarc7_abstract/roomba_request_executer.py]: Also why is the rate limiting being done inside of the state machine?

File: src/iarc7_abstract/roomba_request_executer.py [hunk 4]
```diff
@@ -129,19 +172,46 @@ def _run(cls, roomba_request, status_callback):
...
+                # Waits for the server to finish performing the action.
+                cls._client.wait_for_result()
+                rospy.logwarn("Go to Roomba success: {}".format(
+                    cls._client.get_result()))
+
+            elif state == RoombaRequestExecuterState.FOLLOWING:
+                if not track_sent:
+                    goal = QuadMoveGoal(movement_type="track_roomba",
+                                        frame_id=cls._roomba_id,
+                                        x_overshoot=-.25,
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py:194]: Which frame is this in?

File: src/iarc7_abstract/roomba_request_executer.py [hunk 5]
```diff
@@ -129,19 +172,46 @@ def _run(cls, roomba_request, status_callback):
...
+                                        y_overshoot=-.25)
+                    # Sends the goal to the action server.
+                    cls._client.send_goal(goal)
+                    rospy.logwarn("Following Roomba")
+                    track_sent = True
+
+                if cls._roomba_turning:
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py:201]: Track roomba should be capable of following a roomba after it turns around. So I don't think this check is necessary.

**Comment by amiller27** [comment]
I added some stuff for logging (changed RRE state to a Python enum, so we can get states as strings for debugging, we should also do this for all enums going forward).  The RRE test enters the block state but doesn't land to block the roomba; is this correct behavior?

File: src/iarc7_abstract/roomba_request_executer.py (hunk 3 above)
Comment (aftersomemath) [reply src/iarc7_abstract/roomba_request_executer.py]: I think this is an indicator that the action of bump should be higher level than this state machine.

**Comment by aftersomemath** [comment]
Above is my review from several days ago that I just finished.

**Comment by aftersomemath** [comment]
I did not look at Aaron's commits.

File: src/iarc7_abstract/roomba_request_executer.py (hunk 2 above)
Comment (amiller27) [reply src/iarc7_abstract/roomba_request_executer.py:101]: That's the idea, yes

File: src/iarc7_abstract/roomba_request_executer.py (hunk 4 above)
Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py:194]: the over shoots? its added to the distance between the drone (quad I think) and roomba. Also, looking back, these values maybe should be params.

File: src/iarc7_abstract/roomba_request_executer.py (hunk 5 above)
Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py:201]: the idea is that as soon as the roomba turns around, we transition to blocking the roomba.

File: src/iarc7_abstract/roomba_request_executer.py (hunk 3 above)
Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py]: I actually don't know why I added that. It should not be there. That was a mistake being there. 

As for your point, I am of the opinion that the RRE can be used by a higher level AI that need not worry about the intricacies of bumping a roomba. 

What the RRE is missing is the ability to cancel (or modify) requests after they have been sent. But that functionality is out of the scope of this PR, unless opinion is strong enough that it must be added now.

File: src/iarc7_abstract/roomba_request_executer.py [hunk 6]
```diff
@@ -165,38 +236,64 @@ def _run(cls, roomba_request, status_callback):
...
+                        state = RoombaRequestExecuterState.BLOCK
+                    elif tracking_mode == RoombaRequest.HIT:
+                        state = RoombaRequestExecuterState.HIT
+                    else:
+                        state = RoombaRequestExecuterState.FOLLOWING
+
+                # bump roomba logic
+                elif state == RoombaRequestExecuterState.FOLLOWING:
+                    if cls._roomba_turning:
+                        state = RoombaRequestExecuterState.BLOCK
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py:267]: How is this state left? Is it supposed to be left?

Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py:267]: explained in person

**Review by aftersomemath — DISMISSED** [review]
Was this fully tested? Looks good to me.

File: src/iarc7_abstract/roomba_request_executer.py [hunk 7]
```diff
@@ -89,38 +130,41 @@ def run(cls, roomba_request, status_callback=lambda _: None):
...
-            state = RoombaRequestExecuterState.HIT
-        else:
-            state = RoombaRequestExecuterState.BLOCK
+        state = RoombaRequestExecuterState.GOING_TO
+
+        track_sent = False
 
         status_callback(state)
 
         while not rospy.is_shutdown():
```
Comment (aftersomemath) [inline src/iarc7_abstract/roomba_request_executer.py:164]: This loop does not appear to be rate limited.

Comment (asaba96) [reply src/iarc7_abstract/roomba_request_executer.py:164]: Fixed

**Review by aftersomemath — APPROVED** [review]

**Pull request merged**